File: python/tiershift/log.py
"""Decision log: one JSON line per route. Signals, tier, model, cost, latency. Never message text. Port of src/log.ts."""

from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional

from .types import Decision

DEFAULT_LOG = ".tiershift/decisions.jsonl"

LOG_FIELDS = (
    "ts", "kind", "model", "tier", "requested_tier", "degraded", "fell_back", "confidence", "signals", "code_signals",
    "reason", "est_cost_usd", "cost_usd", "input_tokens", "output_tokens", "jev_latency_ms", "jev_input_tokens", "model_latency_ms",
)


def _now_iso() -> str:
    """ISO 8601 with milliseconds and a Z suffix, like JavaScript's toISOString()."""
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.") + f"{datetime.now(timezone.utc).microsecond // 1000:03d}Z"


def log_entry(path: str, entry: dict[str, Any]) -> None:
    p = Path(path).resolve()
    p.parent.mkdir(parents=True, exist_ok=True)
    with p.open("a", encoding="utf-8") as f:
        f.write(json.dumps(entry, ensure_ascii=False, separators=(",", ":")) + "\n")


def read_log(path: str) -> list[dict[str, Any]]:
    p = Path(path).resolve()
    if not p.exists():
        return []
    return [json.loads(line) for line in p.read_text(encoding="utf-8").splitlines() if line.strip()]


def from_decision(d: Decision, kind: str, **extra: Any) -> dict[str, Any]:
    """Build a log entry. Keys and ordering match the TypeScript writer."""
    entry: dict[str, Any] = {
        "ts": _now_iso(),
        "kind": kind,
        "model": d.model,
        "tier": d.tier,
        "requested_tier": d.requested_tier,
        "degraded": d.degraded,
        "fell_back": False,
        "confidence": d.confidence,
        "signals": d.signals.to_dict(),
        "code_signals": d.code_signals.to_dict(),
        "reason": list(d.reason),
        "est_cost_usd": d.est_cost_usd,
        "cost_usd": None,
        "input_tokens": None,
        "output_tokens": None,
        "jev_latency_ms": d.jev_latency_ms,
        "jev_input_tokens": d.jev_input_tokens,
        "model_latency_ms": None,
    }
    for k, v in extra.items():
        if k == "tag" and v is None:
            continue
        entry[k] = v
    return entry

File: src/log.ts
/** Decision log: one JSON line per route. Signals, tier, model, cost, latency. Never message text. */
import { appendFileSync, existsSync, mkdirSync, readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import type { Decision } from "./types.js";

export interface LogEntry {
  ts: string;
  /** "route" when only a decision was made; "complete" when a model was called. */
  kind: "route" | "complete";
  model: string;
  tier: string;
  requested_tier: string;
  degraded: boolean;
  fell_back: boolean;
  confidence: number;
  signals: Decision["signals"];
  code_signals: Decision["code_signals"];
  reason: string[];
  est_cost_usd: number | null;
  /** Everything billed for the request: failed attempts, gate calls, and the served answer. Null for route-only entries. */
  cost_usd: number | null;
  /** Answer-gate probability for the served answer, when the gate ran. */
  gate_addresses?: number;
  input_tokens: number | null;
  output_tokens: number | null;
  jev_latency_ms: number;
  jev_input_tokens: number;
  model_latency_ms: number | null;
  /** Caller-supplied tag, e.g. an agent name or tenant. */
  tag?: string;
}

export const DEFAULT_LOG = ".tiershift/decisions.jsonl";

export function logEntry(path: string, entry: LogEntry): void {
  const p = resolve(path);
  const dir = dirname(p);
  if (!existsSync(dir)) mkdirSync(dir, { recursive: true });
  appendFileSync(p, JSON.stringify(entry) + "\n");
}

export function readLog(path: string): LogEntry[] {
  const p = resolve(path);
  if (!existsSync(p)) return [];
  return readFileSync(p, "utf8").split("\n").filter((l) => l.trim()).map((l) => JSON.parse(l) as LogEntry);
}

export function fromDecision(d: Decision, extra: Partial<LogEntry> & { kind: LogEntry["kind"] }): LogEntry {
  return {
    ts: new Date().toISOString(),
    model: d.model, tier: d.tier, requested_tier: d.requested_tier, degraded: d.degraded, fell_back: false,
    confidence: d.confidence, signals: d.signals, code_signals: d.code_signals, reason: d.reason,
    est_cost_usd: d.est_cost_usd, cost_usd: null, input_tokens: null, output_tokens: null,
    jev_latency_ms: d.jev_latency_ms, jev_input_tokens: d.jev_input_tokens, model_latency_ms: null,
    ...extra,
  };
}
